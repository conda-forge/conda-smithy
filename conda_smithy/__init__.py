try:
    from conda_smithy._version import __version__
except ImportError:
    __version__ = "0.0.0"
