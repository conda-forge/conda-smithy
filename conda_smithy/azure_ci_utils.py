import os
import typing
import warnings

from msrest.authentication import Authentication, BasicAuthentication
from vsts.build.v4_1.build_client import BuildClient
from vsts.build.v4_1.models import (
    BuildDefinition,
    BuildDefinitionReference,
    SourceRepositories,
    SourceRepository,
)
from vsts.service_endpoint.v4_1.models import ServiceEndpoint
from vsts.service_endpoint.v4_1.service_endpoint_client import (
    ServiceEndpointClient,
)
from vsts.task_agent.v4_0.models import TaskAgentQueue
from vsts.task_agent.v4_0.task_agent_client import TaskAgentClient
from vsts.vss_connection import VssConnection


class AzureConfig:

    _default_org = "conda-forge"
    _default_project_name = "feedstock-builds"

    def __init__(
        self, org_or_user=None, project_name=None, team_instance=None
    ):
        self.org_or_user = org_or_user or os.getenv(
            "AZURE_ORG_OR_USER", self._default_org
        )
        # This will only need to be changed if you need to point to a non-standard azure
        # instance.
        self.instance_base_url = team_instance or os.getenv(
            "AZURE_INSTANCE", f"https://dev.azure.com/{self.org_or_user}"
        )
        self.project_name = project_name or os.getenv(
            "AZURE_PROJECT_NAME", self._default_project_name
        )

        try:
            with open(
                os.path.expanduser("~/.conda-smithy/azure.token"), "r"
            ) as fh:
                self.token = fh.read().strip()
            if not self.token:
                raise ValueError()
        except (IOError, ValueError):
            print(
                "No azure token. Create a token and\n"
                "put it in ~/.conda-smithy/azure.token"
            )
            self.token = None

        # By default for now don't report on the build information back to github
        self.azure_report_build_status = os.getenv(
            "AZURE_REPORT_BUILD_STATUS", "true"
        )

    @property
    def connection(self):
        connection = VssConnection(
            base_url=self.instance_base_url, creds=self.credentials
        )
        return connection

    @property
    def credentials(self):
        if self.token:
            return BasicAuthentication("", self.token)
        else:
            warnings.warn(
                "No token available.  No modifications will be possible!"
            )
            return Authentication()


default_config = AzureConfig()


def get_service_endpoint(config: AzureConfig = default_config):
    service_endpoint_client = ServiceEndpointClient(
        base_url=config.instance_base_url, creds=config.credentials
    )
    endpoints: typing.List[
        ServiceEndpoint
    ] = service_endpoint_client.get_service_endpoints(
        project=config.project_name, type="GitHub"
    )
    for service_endpoint in endpoints:
        if service_endpoint.name == config.org_or_user:
            return service_endpoint
    else:
        raise KeyError("Service endpoint not found")


def get_queues(
    config: AzureConfig = default_config,
) -> typing.List[TaskAgentQueue]:
    aclient = TaskAgentClient(config.instance_base_url, config.credentials)
    return aclient.get_agent_queues(config.project_name)


def get_default_queue(project_name):
    queues = get_queues(project_name)
    for q in queues:
        if q.name == "Default":
            return q
    else:
        raise ValueError("Default queue not found")


def get_repo_reference(config: AzureConfig, github_org, repo_name):
    service_endpoint = get_service_endpoint(config)
    bclient: BuildClient = config.connection.get_client(
        "vsts.build.v4_1.build_client.BuildClient"
    )
    repos: SourceRepositories = bclient.list_repositories(
        project=config.project_name,
        provider_name="github",
        repository=f"{github_org}/{repo_name}",
        service_endpoint_id=service_endpoint.id,
    )

    repo: SourceRepository = repos.repositories[0]
    return repo


def get_default_build_definition(
    github_org, repo_name, config: AzureConfig = default_config, **kwargs,
):
    from vsts.build.v4_1.models import (
        BuildDefinition,
        BuildRepository,
    )
    from vsts.task_agent.v4_0.task_agent_client import TaskAgentClient
    import inspect

    aclient = TaskAgentClient(config.instance_base_url, config.credentials)

    source_repo = get_repo_reference(config, github_org, repo_name)

    new_repo = BuildRepository(
        type="GitHub",
        url=source_repo.properties["cloneUrl"],
        **{
            k: v
            for k, v in source_repo.as_dict().items()
            if k in set(inspect.getfullargspec(BuildRepository).args) - {"url"}
        },
    )
    new_repo.name = source_repo.properties["fullName"]
    new_repo.properties["cleanOptions"] = "0"
    new_repo.properties["skipSyncSource"] = "false"
    new_repo.properties["gitLfsSupport"] = "false"
    new_repo.properties["checkoutNestedSubmodules"] = "false"
    new_repo.properties["labelSources"] = "0"
    new_repo.properties["fetchDepth"] = "0"
    new_repo.properties["labelSourcesFormat"] = "$(build.buildNumber)"
    new_repo.properties["reportBuildStatus"] = config.azure_report_build_status
    new_repo.clean = False

    queues = get_queues(config)
    default_queue = get_default_queue(config)
    service_endpoint = get_service_endpoint(config)

    build_definition = BuildDefinition(
        process={
            "type": 2,
            # These might be optional;
            "resources": {
                "queues": [{"id": q.id, "alias": q.name} for q in queues],
                "endpoints": [
                    {"id": service_endpoint.id, "alias": service_endpoint.name}
                ],
            },
            "yamlFilename": "/azure-pipelines.yml",
        },
        # queue works
        queue=default_queue,
        # now onto this
        repository=new_repo,
        name=repo_name,
        # configure trigger for our builds.
        triggers=[
            {
                "branchFilters": ["+*"],
                "forks": {"enabled": True, "allowSecrets": False},
                "pathFilters": [],
                "isCommentRequiredForPullRequest": False,
                "triggerType": "pullRequest",
            },
            {
                "branchFilters": ["+*"],
                "pathFilters": [],
                "batchChanges": False,
                "maxConcurrentBuildsPerBranch": 1,
                "pollingInterval": 0,
                "triggerType": "continuousIntegration",
            },
        ],
        variable_groups=aclient.get_variable_groups(
            project=config.project_name, group_name="anaconda-org"
        ),
        type="build",
        **kwargs,
    )

    return build_definition


def register_repo(github_org, repo_name, config: AzureConfig = default_config):
    from vsts.build.v4_1.models import BuildDefinitionReference

    bclient = build_client()
    build_definition = get_default_build_definition(
        github_org, repo_name, config=config,
    )

    # clean up existing builds for the same feedstock if present
    existing_definitions: typing.List[
        BuildDefinitionReference
    ] = bclient.get_definitions(project=config.project_name, name=repo_name)
    if existing_definitions:
        assert len(existing_definitions) == 1
        ed = existing_definitions[0]
        bclient.update_definition(
            definition=build_definition,
            definition_id=ed.id,
            project=ed.project.name,
        )
    else:
        bclient.create_definition(
            definition=build_definition, project=config.project_name
        )


def build_client(config: AzureConfig = default_config) -> BuildClient:
    return config.connection.get_client(
        "vsts.build.v4_1.build_client.BuildClient"
    )


def repo_registered(
    github_org: str, repo_name: str, config: AzureConfig = default_config
) -> bool:
    existing_definitions: typing.List[BuildDefinitionReference] = build_client(
        config
    ).get_definitions(project=config.project_name, name=repo_name)

    return bool(existing_definitions)


def enable_reporting(repo, config: AzureConfig = default_config) -> None:
    bclient = build_client(config)
    bdef_header = bclient.get_definitions(
        project=config.project_name, name=repo
    )[0]
    bdef = bclient.get_definition(bdef_header.id, bdef_header.project.name)
    bdef.repository.properties["reportBuildStatus"] = "true"
    bclient.update_definition(bdef, bdef.id, bdef.project.name)


def get_build_id(repo, config: AzureConfig = default_config) -> dict:
    """Get the necessary build information to persist in the config file to allow rendering
    of badges.
    This is needed by non-conda-forge use cases"""
    bclient = build_client(config)
    bdef_header = bclient.get_definitions(
        project=config.project_name, name=repo
    )[0]
    bdef: BuildDefinition = bclient.get_definition(
        bdef_header.id, bdef_header.project.name
    )

    return dict(
        user_or_org=config.org_or_user,
        project_name=config.project_name,
        build_id=bdef.id,
        project_id=bdef.project.id,
    )
