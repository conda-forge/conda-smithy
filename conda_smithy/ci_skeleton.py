"""Generate skeleton for using conda-forge CI configuration files outside of
a feedstock (i.e. a outside of a forge context). This enables people to run
their CI without having to worry about how to set it up or keep it up-to-date,
by reusing the same tools that conda-forge uses for its infrastructure.
Note that your CI jobs will still execute under your organization, and not be
added to conda-forge's queue.
"""
import os
import sys

from jinja2 import Environment, FileSystemLoader

from .configure_feedstock import make_jinja_env, conda_forge_content


def _render_template(template_file, env, forge_dir, config):
    """Renders the template"""
    template = env.get_template(
        os.path.basename(template_file) + ".ci-skel.tmpl"
    )
    target_fname = os.path.join(forge_dir, template_file)
    print("Generating " + target_fname, file=sys.stderr)
    new_file_contents = template.render(**config)
    os.makedirs(os.path.dirname(target_fname), exist_ok=True)
    with open(target_fname, "w") as fh:
        fh.write(new_file_contents)


def _insert_into_gitignore(
    feedstock_directory=".",
    prefix="# conda smithy ci-skeleton start\n",
    suffix="# conda smithy ci-skeleton end\n",
):
    """Places gitignore contents into gitignore."""
    # get current contents
    fname = os.path.join(feedstock_directory, ".gitignore")
    print("Updating " + fname)
    if os.path.isfile(fname):
        with open(fname, "r") as f:
            s = f.read()
        before, _, s = s.partition(prefix)
        _, _, after = s.partition(suffix)
    else:
        before = after = ""
        dname = os.path.dirname(fname)
        if dname:
            os.makedirs(dname, exist_ok=True)
    # get new values
    gi = os.path.join(conda_forge_content, "feedstock_content", ".gitignore")
    with open(gi, "r") as f:
        s = f.read()
    new = prefix + s + suffix
    # write out the file
    with open(fname, "w") as f:
        f.write(before + new + after)
    return fname


def generate(
    package_name="pkg", feedstock_directory=".", recipe_directory="recipe"
):
    """Generates the CI skeleton."""
    forge_dir = os.path.abspath(feedstock_directory)
    env = make_jinja_env(forge_dir)
    config = dict(
        package_name=package_name,
        feedstock_directory=feedstock_directory,
        recipe_directory=recipe_directory,
    )
    # render templates
    _render_template("conda-forge.yml", env, forge_dir, config)
    _render_template(
        os.path.join(recipe_directory, "meta.yaml"), env, forge_dir, config
    )
    # update files which may exist with other content
    _insert_into_gitignore(feedstock_directory=feedstock_directory)
