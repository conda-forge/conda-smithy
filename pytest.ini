[pytest]
markers =
    legacy_circle: Test designed to run as if prior to the azure migration
    legacy_travis: Test designed to run as if prior to the azure migration
    legacy_appveyor: Test designed to run as if prior to the azure migration
    cli: CLI tests outside of test/test_cli.py