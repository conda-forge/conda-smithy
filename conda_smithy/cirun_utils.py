"""
See http://py.cirun.io/api.html for cirun client docs
"""

from functools import lru_cache
from typing import Any, Dict, List, Optional

from cirun import Cirun

from conda_smithy.github import Github, gh_token


@lru_cache
def get_cirun_installation_id(owner: str) -> int:
    # This ID needs a token with admin: org privileges.
    # Hard-code instead for easier use.
    if owner == "conda-forge":
        return 18453316
    else:
        gh = Github(gh_token())
        user = gh.get_user()
        user_or_org: Any
        if user.login == owner:
            user_or_org = user
        else:
            user_or_org = gh.get_organization(owner)
        for inst in user_or_org.get_installations():
            if inst.raw_data["app_slug"] == "cirun-application":
                return inst.app_id
        raise ValueError(f"cirun not found for owner {owner}")


def enable_cirun_for_project(owner: str, repo: str) -> Dict[str, Any]:
    """Enable the cirun.io Github Application for a particular repository."""
    print(f"Enabling cirun for {owner}/{repo} ...")
    cirun = _get_cirun_client()
    return cirun.set_repo(
        f"{owner}/{repo}", installation_id=get_cirun_installation_id(owner)
    )


def add_repo_to_cirun_resource(
    owner: str,
    repo: str,
    resources: List[str],
    teams: Optional[List],
    roles: Optional[List],
    users_from_json: Optional[str] = None,
    cirun_policy_args: Optional[List[str]] = None,
) -> Dict[str, Any]:
    """Grant access to a cirun resource to a particular repository, with a particular policy."""
    cirun = _get_cirun_client()
    policy_args: Optional[Dict[str, Any]] = None
    if cirun_policy_args and "pull_request" in cirun_policy_args:
        policy_args = {"pull_request": True}
    print(
        f"Adding repo {owner}/{repo} to resources {resources} with policy_args: {policy_args}"
    )
    gh = Github(gh_token())
    gh_owner = gh.get_user(owner)
    gh_repo = gh_owner.get_repo(repo)

    # Need to send None instead of an empty list
    teams = teams or None
    roles = roles or None
    if not (teams or roles or users_from_json):
        teams = [team.name for team in gh_repo.get_teams()]

    response = cirun.add_repo_to_resources(
        owner,
        repo,
        resources=resources,
        policy_args=policy_args,
        teams=teams,
        roles=roles,
        users_from_json=users_from_json,
    )
    print(f"response: {response} | {response.json().keys()}")
    return response


def remove_repo_from_cirun_resource(owner: str, repo: str, resource: str):
    """Revoke access to a cirun resource to a particular repository, with a particular policy."""
    cirun = _get_cirun_client()
    print(f"Removing repo {owner}/{repo} from resource {resource}.")
    response = cirun.remove_repo_from_resources(owner, repo, [resource])
    print(f"response: {response} | {response.json().keys()}")
    return response


@lru_cache
def _get_cirun_client() -> Cirun:
    try:
        return Cirun()
    except KeyError:
        raise RuntimeError(
            "You must have CIRUN_API_KEY defined to do Cirun CI registration. "
            "This requirement can be overriden by specifying `--without-cirun`"
        )
