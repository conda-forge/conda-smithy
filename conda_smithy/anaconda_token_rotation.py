"""This module updates/rotates anaconda/binstar tokens.

The correct way to use this module is to call its functions via the command
line utility. The relevant one is

    conda-smithy update-anaconda-token

Note that if you are using appveyor, you will need to push the changes to the
conda-forge.yml in your feedstock to GitHub.
"""

import os
import sys
from contextlib import redirect_stderr, redirect_stdout
from typing import Optional, Union

import requests
from github import Github

from conda_smithy.utils import update_conda_forge_config


def _get_anaconda_token():
    """use this helper to enable easier patching for tests"""
    try:
        from conda_smithy.ci_register import anaconda_token

        return anaconda_token
    except ImportError:
        raise RuntimeError(
            "You must have the anaconda token defined to do token rotation!"
        )


def rotate_anaconda_token(
    user: str,
    project: str,
    feedstock_config_path: Optional[str],
    drone: bool = True,
    circle: bool = True,
    travis: bool = True,
    azure: bool = True,
    appveyor: bool = True,
    github_actions: bool = True,
    token_name: str = "BINSTAR_TOKEN",
    drone_endpoints: Union[list, tuple] = (),
):
    """Rotate the anaconda (binstar) token used by the CI providers

    All exceptions are swallowed and stdout/stderr from this function is
    redirected to `/dev/null`. Sanitized error messages are
    displayed at the end.

    If you need to debug this function, define `DEBUG_ANACONDA_TOKENS` in
    your environment before calling this function.
    """
    # we are swallong all of the logs below, so we do a test import here
    # to generate the proper errors for missing tokens
    # note that these imports cover all providers
    from .ci_register import travis_endpoint  # noqa
    from .azure_ci_utils import default_config  # noqa
    from conda_smithy.github import gh_token

    anaconda_token = _get_anaconda_token()

    if github_actions:
        gh = Github(gh_token())

    # capture stdout, stderr and suppress all exceptions so we don't
    # spill tokens
    failed = False
    err_msg = None
    with open(os.devnull, "w") as fp:
        if "DEBUG_ANACONDA_TOKENS" in os.environ:
            fpo = sys.stdout
            fpe = sys.stderr
        else:
            fpo = fp
            fpe = fp

        with redirect_stdout(fpo), redirect_stderr(fpe):
            try:
                if circle:
                    try:
                        rotate_token_in_circle(
                            user, project, anaconda_token, token_name
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on circle!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if drone:
                    for drone_endpoint in drone_endpoints:
                        try:
                            rotate_token_in_drone(
                                user,
                                project,
                                anaconda_token,
                                token_name,
                                drone_endpoint,
                            )
                        except Exception as e:
                            if "DEBUG_ANACONDA_TOKENS" in os.environ:
                                raise e
                            else:
                                err_msg = (
                                    f"Failed to rotate token for {user}/{project}"
                                    f" on drone endpoint {drone_endpoint}!"
                                )
                                failed = True
                                raise RuntimeError(err_msg)

                if travis:
                    try:
                        rotate_token_in_travis(
                            user,
                            project,
                            feedstock_config_path,
                            anaconda_token,
                            token_name,
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on travis!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if azure:
                    try:
                        rotate_token_in_azure(
                            user, project, anaconda_token, token_name
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on azure!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if appveyor:
                    try:
                        rotate_token_in_appveyor(
                            feedstock_config_path, anaconda_token, token_name
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on appveyor!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

                if github_actions:
                    try:
                        rotate_token_in_github_actions(
                            user, project, anaconda_token, token_name, gh
                        )
                    except Exception as e:
                        if "DEBUG_ANACONDA_TOKENS" in os.environ:
                            raise e
                        else:
                            err_msg = (
                                f"Failed to rotate token for {user}/{project}"
                                " on github actions!"
                            )
                            failed = True
                            raise RuntimeError(err_msg)

            except Exception as e:
                if "DEBUG_ANACONDA_TOKENS" in os.environ:
                    raise e
                failed = True

    if failed:
        if err_msg:
            raise RuntimeError(err_msg)
        else:
            raise RuntimeError(
                f"Rotating the feedstock token in providers for {user}/{project} failed!"
                " Try the command locally with DEBUG_ANACONDA_TOKENS"
                " defined in the environment to investigate!"
            )


def rotate_token_in_circle(user, project, binstar_token, token_name):
    from conda_smithy.ci_register import circle_token

    url_template = (
        "https://circleci.com/api/v1.1/project/github/{user}/{project}/envvar{extra}?"
        "circle-token={token}"
    )

    r = requests.get(
        url_template.format(
            token=circle_token, user=user, project=project, extra=""
        )
    )
    if r.status_code != 200:
        r.raise_for_status()

    have_binstar_token = False
    for evar in r.json():
        if evar["name"] == token_name:
            have_binstar_token = True

    if have_binstar_token:
        r = requests.delete(
            url_template.format(
                token=circle_token,
                user=user,
                project=project,
                extra=f"/{token_name}",
            )
        )
        if r.status_code != 200:
            r.raise_for_status()

    data = {"name": token_name, "value": binstar_token}
    response = requests.post(
        url_template.format(
            token=circle_token, user=user, project=project, extra=""
        ),
        data,
    )
    if response.status_code != 201:
        raise ValueError(response)


def rotate_token_in_drone(
    user, project, binstar_token, token_name, drone_endpoint
):
    from conda_smithy.ci_register import drone_session

    session = drone_session(drone_endpoint)

    r = session.get(f"/api/repos/{user}/{project}/secrets")
    r.raise_for_status()
    have_binstar_token = False
    for secret in r.json():
        if token_name == secret["name"]:
            have_binstar_token = True

    if have_binstar_token:
        r = session.patch(
            f"/api/repos/{user}/{project}/secrets/{token_name}",
            json={"data": binstar_token, "pull_request": False},
        )
        r.raise_for_status()
    else:
        response = session.post(
            f"/api/repos/{user}/{project}/secrets",
            json={
                "name": token_name,
                "data": binstar_token,
                "pull_request": False,
            },
        )
        if response.status_code != 200:
            response.raise_for_status()


def rotate_token_in_travis(
    user, project, feedstock_config_path, binstar_token, token_name
):
    """update the binstar token in travis."""
    from conda_smithy.ci_register import (
        travis_endpoint,
        travis_get_repo_info,
        travis_headers,
    )

    headers = travis_headers()

    repo_info = travis_get_repo_info(user, project)
    repo_id = repo_info["id"]

    r = requests.get(
        f"{travis_endpoint}/repo/{repo_id}/env_vars",
        headers=headers,
    )
    if r.status_code != 200:
        r.raise_for_status()

    have_binstar_token = False
    ev_id = None
    for ev in r.json()["env_vars"]:
        if ev["name"] == token_name:
            have_binstar_token = True
            ev_id = ev["id"]

    data = {
        "env_var.name": token_name,
        "env_var.value": binstar_token,
        "env_var.public": "false",
    }

    if have_binstar_token:
        r = requests.patch(
            f"{travis_endpoint}/repo/{repo_id}/env_var/{ev_id}",
            headers=headers,
            json=data,
        )
        r.raise_for_status()
    else:
        r = requests.post(
            f"{travis_endpoint}/repo/{repo_id}/env_vars",
            headers=headers,
            json=data,
        )
        if r.status_code != 201:
            r.raise_for_status()

    # we remove the token in the conda-forge.yml since on travis the
    # encrypted values override any value we put in the API
    with update_conda_forge_config(feedstock_config_path) as code:
        if (
            "travis" in code
            and "secure" in code["travis"]
            and token_name in code["travis"]["secure"]
        ):
            del code["travis"]["secure"][token_name]

            if len(code["travis"]["secure"]) == 0:
                del code["travis"]["secure"]

            if len(code["travis"]) == 0:
                del code["travis"]

            print(
                "An old value of the variable %s for travis was found in the "
                "conda-forge.yml. You may need to rerender this feedstock to "
                "use the new value since encrypted secrets inserted in travis.yml "
                "files override those set in the UI/API!"
            )


def rotate_token_in_azure(user, project, binstar_token, token_name):
    from vsts.build.v4_1.models import BuildDefinitionVariable

    from conda_smithy.azure_ci_utils import (
        build_client,
        get_default_build_definition,
    )
    from conda_smithy.azure_ci_utils import default_config as config

    bclient = build_client()

    existing_definitions = bclient.get_definitions(
        project=config.project_name, name=project
    )
    if existing_definitions:
        assert len(existing_definitions) == 1
        ed = existing_definitions[0]
    else:
        raise RuntimeError(
            f"Cannot add {token_name} to a repo that is not already registerd on azure CI!"
        )

    ed = bclient.get_definition(ed.id, project=config.project_name)

    if not hasattr(ed, "variables") or ed.variables is None:
        variables = {}
    else:
        variables = ed.variables

    variables[token_name] = BuildDefinitionVariable(
        allow_override=False,
        is_secret=True,
        value=binstar_token,
    )

    build_definition = get_default_build_definition(
        user,
        project,
        config=config,
        variables=variables,
        id=ed.id,
        revision=ed.revision,
    )

    bclient.update_definition(
        definition=build_definition,
        definition_id=ed.id,
        project=ed.project.name,
    )


def rotate_token_in_appveyor(feedstock_config_path, binstar_token, token_name):
    from conda_smithy.ci_register import appveyor_token

    headers = {"Authorization": f"Bearer {appveyor_token}"}
    url = "https://ci.appveyor.com/api/account/encrypt"
    response = requests.post(
        url, headers=headers, data={"plainValue": binstar_token}
    )
    if response.status_code != 200:
        raise ValueError(response)

    with update_conda_forge_config(feedstock_config_path) as code:
        code.setdefault("appveyor", {}).setdefault("secure", {})[
            token_name
        ] = response.content.decode("utf-8")


def rotate_token_in_github_actions(
    user, project, binstar_token, token_name, gh
):
    repo = gh.get_repo(f"{user}/{project}")
    assert repo.create_secret(token_name, binstar_token)
