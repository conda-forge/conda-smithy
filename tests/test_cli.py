import argparse
import collections
import os
import subprocess
import yaml

from conda_smithy import cli

_thisdir = os.path.abspath(os.path.dirname(__file__))

InitArgs = collections.namedtuple('ArgsObject', ('no_git_repo',
                                                 'recipe_directory',
                                                 'feedstock_directory',
                                                 'variant_config_files'))

RegenerateArgs = collections.namedtuple('ArgsObject', ('commit',
                                                       'feedstock_directory',
                                                       'variant_config_files'))


def test_init(py_recipe):
    """This is the command that takes the initial staged-recipe folder and turns it into a
    feedstock"""
    # actual parser doesn't matter.  It's used for initialization only
    parser = argparse.ArgumentParser()
    subparser = parser.add_subparsers()
    init_obj = cli.Init(subparser)
    recipe = py_recipe.recipe
    # expected args object has
    args = InitArgs(recipe_directory=os.path.join(recipe, 'recipe'),
                    feedstock_directory=os.path.join(recipe, '{package.name}-feedstock'),
                    no_git_repo=False,
                    variant_config_files=os.path.join(recipe, 'config.yaml'))
    init_obj(args)
    destination = os.path.join(recipe, 'py-test-feedstock')
    assert os.path.isdir(destination)


#TODO: remove the 2 lines below. https://github.com/conda-forge/conda-smithy/issues/650
import pytest
@pytest.mark.xfail
def test_init_multiple_output_matrix(testing_workdir):
    parser = argparse.ArgumentParser()
    subparser = parser.add_subparsers()
    init_obj = cli.Init(subparser)
    recipe = os.path.join(_thisdir, 'recipes', 'multiple_outputs')
    feedstock_dir = os.path.join(testing_workdir, 'multiple-outputs-test-feedstock')
    args = InitArgs(recipe_directory=recipe,
                    feedstock_directory=feedstock_dir,
                    no_git_repo=False,
                    variant_config_files=os.path.join(recipe, 'conda_build_config.yaml'))
    init_obj(args)
    matrix_dir = os.path.join(feedstock_dir, '.ci_support')
    # the matrix should be consolidated among all outputs, as well as the top-level
    # reqs. Only the top-level reqs should have indedependent config files,
    # though - loops within outputs are contained in those top-level configs.
    assert len(os.listdir(matrix_dir)) == 13
    circle_libpng16 = os.path.join(matrix_dir, 'circle_libpng1.6libpq9.5.yaml')
    assert os.path.isfile(circle_libpng16)
    with open(circle_libpng16) as f:
        config = yaml.load(f)
    assert config['libpng'] == ['1.6']
    assert config['libpq'] == ['9.5']
    # this is a zipped key, but it's not used, so it shouldn't show up
    assert 'libtiff' not in config
    assert 'zip_keys' not in config or not any('libtiff' in group for group in config['zip_keys'])
    # this is a variable only for one of the outputs
    assert config['jpeg'] == ['8', '9']
    # this is in conda_build_config.yaml, but is a transitive dependency.  It should
    #     not show up in the final configs.
    assert 'zlib' not in config


def test_regenerate(py_recipe):
    parser = argparse.ArgumentParser()
    subparser = parser.add_subparsers()
    regen_obj = cli.Regenerate(subparser)
    recipe = py_recipe.recipe
    feedstock_dir = os.path.join(_thisdir, 'recipes', 'click-test-feedstock')
    args = RegenerateArgs(feedstock_directory=feedstock_dir,
                          commit=False,
                          variant_config_files=os.path.join(recipe, 'config.yaml'))
    matrix_folder = os.path.join(feedstock_dir, '.ci_support')

    initial_commit = subprocess.check_output(['git', 'rev-parse', 'HEAD'],
                                             cwd=feedstock_dir).strip()
    try:
        # original rendering was with py27, 36, no target_platform
        assert len(os.listdir(matrix_folder)) == 7
        regen_obj(args)
        # should add 2, as the config.yaml adds in target_platform
        assert len(os.listdir(matrix_folder)) == 9

        # reduce the python matrix and make sure the matrix files reflect the change
        args = RegenerateArgs(feedstock_directory=feedstock_dir,
                              commit=False,
                              variant_config_files=os.path.join(recipe, 'short_config.yaml'))
        matrix_folder = os.path.join(feedstock_dir, '.ci_support')
        # one py ver, no target_platform  (tests that older configs don't stick around)
        regen_obj(args)
        assert len(os.listdir(matrix_folder)) == 4
    finally:
        # reset the test dir for next time
        subprocess.call(['git', 'reset', '--hard', initial_commit],
                        cwd=feedstock_dir)
