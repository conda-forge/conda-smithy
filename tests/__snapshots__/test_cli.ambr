# serializer version: 1
# name: test_init_cuda_docker_images
  dict({
    'cdt_name': list([
      'cos6',
    ]),
    'cuda_compiler': list([
      'nvcc',
    ]),
    'cuda_compiler_version': list([
      'None',
    ]),
    'docker_image': list([
      'condaforge/linux-anvil-comp7',
    ]),
    'target_platform': list([
      'linux-64',
    ]),
    'zip_keys': list([
      list([
        'cdt_name',
        'cuda_compiler_version',
        'docker_image',
      ]),
    ]),
  })
# ---
# name: test_init_cuda_docker_images[10.0]
  dict({
    'cdt_name': list([
      'cos6',
    ]),
    'cuda_compiler': list([
      'nvcc',
    ]),
    'cuda_compiler_version': list([
      '10.0',
    ]),
    'docker_image': list([
      'condaforge/linux-anvil-cuda:10.0',
    ]),
    'target_platform': list([
      'linux-64',
    ]),
    'zip_keys': list([
      list([
        'cdt_name',
        'cuda_compiler_version',
        'docker_image',
      ]),
    ]),
  })
# ---
# name: test_init_cuda_docker_images[10.1]
  dict({
    'cdt_name': list([
      'cos6',
    ]),
    'cuda_compiler': list([
      'nvcc',
    ]),
    'cuda_compiler_version': list([
      '10.1',
    ]),
    'docker_image': list([
      'condaforge/linux-anvil-cuda:10.1',
    ]),
    'target_platform': list([
      'linux-64',
    ]),
    'zip_keys': list([
      list([
        'cdt_name',
        'cuda_compiler_version',
        'docker_image',
      ]),
    ]),
  })
# ---
# name: test_init_cuda_docker_images[10.2]
  dict({
    'cdt_name': list([
      'cos6',
    ]),
    'cuda_compiler': list([
      'nvcc',
    ]),
    'cuda_compiler_version': list([
      '10.2',
    ]),
    'docker_image': list([
      'condaforge/linux-anvil-cuda:10.2',
    ]),
    'target_platform': list([
      'linux-64',
    ]),
    'zip_keys': list([
      list([
        'cdt_name',
        'cuda_compiler_version',
        'docker_image',
      ]),
    ]),
  })
# ---
# name: test_init_cuda_docker_images[11.0]
  dict({
    'cdt_name': list([
      'cos7',
    ]),
    'cuda_compiler': list([
      'nvcc',
    ]),
    'cuda_compiler_version': list([
      '11.0',
    ]),
    'docker_image': list([
      'condaforge/linux-anvil-cuda:11.0',
    ]),
    'target_platform': list([
      'linux-64',
    ]),
    'zip_keys': list([
      list([
        'cdt_name',
        'cuda_compiler_version',
        'docker_image',
      ]),
    ]),
  })
# ---
# name: test_init_cuda_docker_images[9.2]
  dict({
    'cdt_name': list([
      'cos6',
    ]),
    'cuda_compiler': list([
      'nvcc',
    ]),
    'cuda_compiler_version': list([
      '9.2',
    ]),
    'docker_image': list([
      'condaforge/linux-anvil-cuda:9.2',
    ]),
    'target_platform': list([
      'linux-64',
    ]),
    'zip_keys': list([
      list([
        'cdt_name',
        'cuda_compiler_version',
        'docker_image',
      ]),
    ]),
  })
# ---
# name: test_init_multiple_docker_images
  dict({
    'cdt_name': list([
      'pickme_1',
    ]),
    'docker_image': list([
      'pickme_a',
    ]),
    'target_platform': list([
      'linux-64',
    ]),
    'zip_keys': list([
      list([
        'docker_image',
        'cdt_name',
      ]),
    ]),
  })
# ---
# name: test_init_with_custom_config[conda-build]
  dict({
    'bot': dict({
      'automerge': True,
      'run_deps_from_wheel': True,
    }),
  })
# ---
# name: test_init_with_custom_config[rattler-build]
  dict({
    'bot': dict({
      'automerge': True,
      'run_deps_from_wheel': True,
    }),
    'conda_build_tool': 'rattler-build',
  })
# ---
